File: backend/src/main/java/tr/com/targe/iot/service/SensorValuesService.java
package tr.com.targe.iot.service;

import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import tr.com.targe.iot.repository.SensorValuesRepository;
import tr.com.targe.iot.entity.SensorValues;
import tr.com.targe.iot.mapper.SensorValuesMapper;
import java.util.List;
import java.util.stream.Collectors;
import tr.com.targe.iot.DTO.SensorValuesDTO;

@Service
@RequiredArgsConstructor
@Transactional
public class SensorValuesService {
    
    private final SensorValuesRepository sensorValuesRepository;
    private final SensorValuesMapper sensorValuesMapper;

    // Tüm sensör değerlerini getir
    public List<SensorValuesDTO> getAllSensorValues() {
        List<SensorValues> sensorValues = sensorValuesRepository.findAll();
        return sensorValues.stream()
            .map(sensorValuesMapper::toDTO)
            .collect(Collectors.toList());
    }

    // Belirli bir cihazın tüm sensör verilerini getir
    public List<SensorValuesDTO> getSensorValuesByDeviceId(Long deviceId) {
        List<SensorValues> sensorValues = sensorValuesRepository.findByDeviceId(deviceId);
        return sensorValues.stream()
            .map(sensorValuesMapper::toDTO)
            .collect(Collectors.toList());
    }

    // Her cihaz için en son sensör verilerini getir
    public List<SensorValuesDTO> getLatestValuesForAllDevices() {
        List<SensorValues> latestValues = sensorValuesRepository.findLatestValuesForAllDevices();
        return latestValues.stream()
            .map(sensorValuesMapper::toDTO)
            .collect(Collectors.toList());
    }

    // Belirli bir cihazın en son sensör verilerini getir
    public List<SensorValuesDTO> getLatestValuesForDevice(Long deviceId) {
        List<SensorValues> latestValues = sensorValuesRepository.findLatestValuesForDevice(deviceId);
        return latestValues.stream()
            .map(sensorValuesMapper::toDTO)
            .collect(Collectors.toList());
    }


    // Son N kayıt için cihaz bazlı sensör verilerini getir
    public List<SensorValuesDTO> getLastNValuesByDevice(Long deviceId, int count) {
        List<SensorValues> lastValues = sensorValuesRepository.findLastNValuesByDevice(
            deviceId, PageRequest.of(0, count));
        return lastValues.stream()
            .map(sensorValuesMapper::toDTO)
            .collect(Collectors.toList());
    }
}

File: frontend/iot-system-management-frontend/src/services/sensorValuesService.js
import axios from "axios";

const BASE_URL = "http://localhost:8080/api/sensor-values";

const sensorValuesService = {
  // Tüm sensör değerlerini getir
  getAllSensorValues: async () => {
    try {
      const response = await axios.get(BASE_URL);
      return response.data;
    } catch (error) {
      console.error("Error fetching all sensor values:", error);
      throw error;
    }
  },

  // Belirli bir cihazın tüm sensör değerlerini getir
  getSensorValuesByDeviceId: async (deviceId) => {
    try {
      const response = await axios.get(`${BASE_URL}/device/${deviceId}`);
      return response.data;
    } catch (error) {
      console.error("Error fetching sensor values by device ID:", error);
      throw error;
    }
  },
  // Belirli bir cihazın en son sensör değerlerini getir
  getLatestValuesForDevice: async (deviceId) => {
    try {
      const response = await axios.get(`${BASE_URL}/latest/device/${deviceId}`);
      return response.data;
    } catch (error) {
      console.error("Error fetching latest values for device:", error);
      throw error;
    }
  },

  // Tüm cihazların en son sensör değerlerini getir
  getLatestValuesForAllDevices: async () => {
    try {
      const response = await axios.get(`${BASE_URL}/latest`);
      return response.data;
    } catch (error) {
      console.error("Error fetching latest values for all devices:", error);
      throw error;
    }
  },

  // Belirli bir sensör tipi için tüm cihazların en son değerlerini getir
  getLatestValuesBySensorType: async (sensorType) => {
    try {
      const response = await axios.get(`${BASE_URL}/latest/type/${sensorType}`);
      return response.data;
    } catch (error) {
      console.error("Error fetching latest values by sensor type:", error);
      throw error;
    }
  },

  // Belirli bir cihazın belirli bir sensör tipi için son N değerini getir
  getLastNValuesByDeviceAndType: async (deviceId, sensorType, count) => {
    try {
      const response = await axios.get(
        `${BASE_URL}/device/${deviceId}/type/${sensorType}/last/${count}`
      );
      return response.data;
    } catch (error) {
      console.error("Error fetching last N values:", error);
      throw error;
    }
  },
};

export default sensorValuesService;
